updates for koa2.x, use compose for stacks

jbielick opened the pull request:
Opening this for discussion and thoughts:

I was messing with Koa 2 alpha 2 and needed a router — breaking changes around passing `ctx` as an argument yields `koa-router` unusable right now. There's another project [`ko66`](https://github.com/menems/koa-66) that seems to be a similar project that's been setup for Koa 2 support, but I'd like to see this lib grow and I'm sure others would as well.

This PR updates all middleware to the Promise-based Koa 2 async pattern. It does not use `async/await`notation or transpiling, which I could see being an options for this lib as well. For simplicity's sake, everything `GeneratorFunction` was changed to a function with promise-based flow, all middleware functions were updates to the `(ctx, next)` signature and `ctx` is now used in place of `this` when modifying the middleware context. 

Additionally, `koa-compose` was added as a dependency to drive layer middleware stacks. I think this is a better choice than invoking the layer chain of middleware functions by hand (ordered). Since `koa-compose` drives the actual koa middleware execution, there's likely going to be fewer issues when koa middleware api changes, as the compose lib can simply be updated. This also allows for exact same async flow in normal koa middleware due to the same dispatch function driving https://github.com/koajs/compose/blob/master/index.js#L31

By using Koa 2, >= node 4 would be *required*
> (Koa requires node v4.0.0 or higher for (partial) ES2015 support.)

I have not updated the `history.md` with these notes.

### Updates:

* update koa dependency to alpha2
* add koa-compose dependency for layer stack composition
* update middleware signature to `(ctx, next)`
* update param middleware signature to `(param, ctx, next)`
* remove all generator functions and gen type checks
* update all specs for promise-based usage
* updates examples / comments / docs

felixfbecker commented:
Wow. I tried to do a PR too and got stuck when I saw that the stack is composed without koa-compose currently. I'm not 100% confident with koa-66 yet so @alexmingoia could merge this it would be great.

jbielick commented:
Haha. Thanks for checking it out, @felixfbecker . That was a tricky part. I thought you were doing something similar because I also had issues with regenerator "missing ) after argument" error and referenced your babel gh issue to help me fix it :D

jbielick commented on lib/router.js:
```diff
@@ -295,43 +298,28 @@ Router.prototype.prefix = function (prefix) {
-    }
+    layerChain = matched.pathAndMethod.reduce(function(memo, layer) {
+      ctx.captures = layer.captures(path, ctx.captures);
+      ctx.params = layer.params(path, ctx.captures, ctx.params);
```
I'm not sure I implemented this correctly. Should `ctx.captures` and `ctx.params` be re-set for every layer before it's executed?

jbielick replied:
The only way I could see it working would be wrapping each layer and creating more closures but that sounds suboptimal for something that occurs JIT. Maybe these stacks should be pre-composed during `middleware()` or something?

ide replied:
Could you port the old logic without koa-compose?

jbielick replied:
Fair question, but are you familiar with the API changes in Koa between 1.x and 2.x?

Middleware is driven in Koa by Koa-compose. By rolling your own middleware driver you're rewriting Koa-compose and creating inconsistencies in how asynchronous flow is achieved through the stack. Take a look at Koa-compose's dispatch method; that's not something a Koa-compatible router should be re-implementing

jbielick replied:
If this lib used koa-compose from the start, this migration would be trivial. As can be seen by koa66—it's basically a fork of this that'll get more traction by writing less code (using koa-compose).

ide replied:
Unifying the implementations would be nice. I guess I was wondering if borrowing the 10-15 lines of code and adjusting them to work with layers would work and let kia-router v6 ship since it's one of the biggest blockers of koa2. Could then make changes to koa-compose upstream to accommodate koa-router's use case and then clean up the bit of code here.

jbielick commented:
Updated the changes @ https://github.com/jbielick/koa-router/commit/9e742f1153a04d6cefa80fc550de62f62c0a8721#diff-504fd89f344be1e9c4b583f3ebd7428aR319 to reflect the intended behavior. Prepends the layer stack of middleware with one middleware function that supplies params and captures to the context before yielding to the next middleware. 

This essentially flattens the layers and applies the param-builders before each layer stack.

```
    layer1                    layer2
        |                             |
parambuilder        parambuilder
      /    \                       /     \
 mw1  mw2            mw3   mw4 
```

would become

```
[layer1:parambuilder, mw1, m2, layer2:parambuilder, mw3, mw4]
```

Like the original code which iterated over each layer, modified ctx and then invoked each middleware in the layer before continuing to the next layer and doing the same.

alexmingoia commented:
Wow this is great. I will have time to review later today and I'll try to get this merged ASAP.
